DESIGN.md

```diff
@@ -0,0 +1,98 @@
+   to keep these files? Indefinitely is going to be very costly. A year would be
+   cheaper, but completely arbitrary. Keeping just the last L0 and the
+   associated L5 files would require some code, but would probably be the
+   minimum we could keep to enable a restore from a blank slate.
+-  Thorough testing of every part of the process before merging any pull
+   requests.
+-  Separate, managed accounts for infrastructure testing and production.
+-  Trivial deployment.
+-  Monitor performance characteristics.
+-  Verbose logging to central ElasticSearch instance.
+-  Alerts:
+   -  Process start/finish with result and auxiliary information in case of
+      failure.
+   -  Big subprocesses' start/finish.
+
+## Wanted features
+
+-  _Pull rather than push data from Landonline._ This way we can keep the entire
+   system together, rather than having a costly and awkward split between
+   on-premise and cloud architecture.
+-  _Continuous deployment._ Any time a change is merged to the main branch some
+   smoke tests are performed and the result is deployed to production.
+-  _React to change rather than run on a timer,_ where possible. Ideally the BDE
+   Processor should know about new files as soon as the last file has been
+   closed for writing, and should immediately start processing.
+-  _From-scratch deployment friendly._ That is, we should always have the option
+   to deploy from scratch, deleting _all_ the resources we've created and
+   getting back to an identical state (from the end users' perspective) in a
+   timely fashion (that is, before the next data dump).
+-  _Access for everybody at the org._ This should be simpler than the current
+   solution, and should be easy with read replicas.
+-  _Public access,_ which in turn requires:
+   -  A separate user with read-only access to only the appropriately licensed
+      tables (some have a restrictive customer license).
+   -  Some way to
+      [make sure users know about the Creative Commons licensing of the content](https://github.com/linz/bde-processor/pull/9#discussion_r1064211308),
+      maybe by having to get an API key to access it.
+-  _Minimal code changes for upstream schema changes._ For example, rather than
+   using complex schema migration code, we should be able to just change the ETL
+   code and redeploy.
+-  Read replica use _does not affect the performance of the pipeline to LDS._
+   Basically just means we'll need to use separate read replicas for them.
+-  Zero configuration. The service should rely on Azure SSO so we don't need to
+   maintain users, we should not need to do any IP whitelisting, and the current
+   version of the BDE Processor should only support the current version of
+   Landonline.
+
+## Non-requirements
+
+-  _No table versioning._ It creates a _lot_ of extra complexity, and we don't
+   have any PostgreSQL experts at hand to maintain the extensions.
+   -  **TBD:** Review how versioned tables are currently being used, and work
```

Comment:
Interesting! We should check with Positioning and Koordinates whether these use cases are still valid, but for now I've moved it to a user requirement.

What is the "Koordinates import process"? Do they read our DB to produce the next L5 BDE file?

Reply:
Koordinates reads our database to detect changes and then updates their copy of the data (inside the Koordinates / LDS SaaS system) to provide the data to customers via the LDS. We could check with them if it is still valid though, or if there is another way. They can import data from postgres databases without table versioning for example, but typically on smaller datasets probably. It can take hours (days?) for them to scan and update data.

Reply:
I don't think the LDS BDE Loader is used - a quick [search](https://github.com/search?p=1&q=org%3Alinz+lds-bde-loader&type=Code) shows only a single reference outside the repo, and that's just some (very likely obsolete) documentation.